### PR TITLE
Added Greengrass discovery client and sample

### PR DESCRIPTION
**Pull request by DavidOgunsAWS**
* Added client to call Greengrass Discover API: https://docs.aws.amazon.com/greengrass/latest/developerguide/gg-discover-api.html
 * Added sample app project to use API and turn around and utilize discovered core and endpoint configuration to perform basic MQTT operations.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+            final Optional<GGGroup> groupOpt = response.getGGGroups().stream().findFirst();
+            if(groupOpt.isPresent()) {
+                final GGGroup group = groupOpt.get();
+                final GGCore core = group.getCores().stream().findFirst().get();
+                final String dnsOrIp = core.getConnectivity().get(0).getHostAddress();
+                final Integer port = core.getConnectivity().get(0).getPortNumber();
+                System.out.println(String.format("Connecting to group ID %s, with thing arn %s, using endpoint %s:%d",
+                        group.getGGGroupId(), core.getThingArn(), dnsOrIp, port));
+
+                final AwsIotMqttConnectionBuilder connectionBuilder = AwsIotMqttConnectionBuilder.newMtlsBuilderFromPath(certPath, keyPath)
```
Comment (bretambrose) [inline samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java:261]: This should be try-blocked and the returned connection be assigned within a try-block I think.

File: samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                    Integer result = publishResult.get();
+                    System.out.println("Publish result code: " + result);
+                }
+            }
+        } catch (CrtRuntimeException | InterruptedException | ExecutionException ex) {
+            System.out.println("Exception encountered: " + ex.toString());
+            ex.printStackTrace();
+        }
+        finally {
+            if(connection != null && !connection.isNull()) {
```
Comment (bretambrose) [inline samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java:235]: With proper trys, you shouldn't have to do this, you can just let the connection disconnect happen at the end of the block.

File: samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                CompletableFuture<Void> disconnected = connection.disconnect();
+                try {
+                    disconnected.get();
+                } catch (ExecutionException | InterruptedException e) {
+                    throw new RuntimeException("Problem disconnecting the MQTT connection cleanly: " + e.getMessage());
+                }
+            }
+        }
+        System.out.println("Complete!");
+    }
```
Comment (bretambrose) [inline samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java:245]: Similar to Shadow or PubSubStress, add a CrtResource.waitForNoResources(); at the end here and then run the sample using CRT debugging (-Daws.crt.debugnative) and to track down any leaks you'll probably want to enable logging using system properties (https://github.com/awslabs/aws-crt-java/blob/master/src/main/java/software/amazon/awssdk/crt/Log.java#L32)

File: samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            if(rootCaPath != null) {
+                tlsCtxOptions.overrideDefaultTrustStoreFromPath(null, rootCaPath);
+            }
+            HttpProxyOptions proxyOptions = null;
+            if (proxyHost != null && proxyPort > 0) {
+                proxyOptions = new HttpProxyOptions();
+                proxyOptions.setHost(proxyHost);
+                proxyOptions.setPort(proxyPort);
+            }
+            final TlsContext tlsCtx = new TlsContext(tlsCtxOptions);
```
Comment (bretambrose) [inline samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java:189]: tlsCtx and discoveryClient should also be try-block protected

File: samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.Scanner;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static software.amazon.awssdk.iot.discovery.DiscoveryClient.TLS_EXT_ALPN;
+
+class Discovery {
```
[the remark this answers is not in the document]
Comment (DavidOgunsAWS) [reply samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java:39]: This is also an accident. javac more recently seems to not care what the file name is and will still compile and output a class file using the actual class name. This is still confusing so I will fix.

File: sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                                    responseInfo.put(header.getName(), header.getValue());
+                                });
+                            }
+                            @Override
+                            public int onResponseBody(HttpStream stream, byte bodyBytes[]) {
+                                jsonBodyResponseBuilder.append(new String(bodyBytes, StandardCharsets.UTF_8));
+                                return bodyBytes.length;
+                            }
+                            @Override
+                            public void onResponseComplete(HttpStream httpStream, int i) {
```
[the remark this answers is not in the document]
Comment (DavidOgunsAWS) [reply sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java:74]: IDE autogenerated! Will correct along with others

File: sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.StringReader;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+public class DiscoveryClient {
```
[the remark this answers is not in the document]
Comment (DavidOgunsAWS) [reply sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java:17]: Decided this should be Autocloseable. The internally managed HttpClientConnectionManager is a CrtResource which needs to cleaned up so this object needs an interface to trigger that as well.

**Pull request merged**